File: TypeScript/algorithms/searching/linearsearch.ts
/* Linear Search

Description: Linear Search finds the position of a target element in an array by checking each element sequentially from the beginning until the target is found or the end of the array is reached.

Time Complexity: O(n)
Space Complexity: O(1)
*/

function linearSearch<T>(arr: T[], target: T): number {
  if (!Array.isArray(arr)) {
    throw new Error("First argument must be an array");
  }

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === target) {
      return i;
    }
  }
  return -1;
}

function runTests(): void {
  console.log("---------- Linear Search Test Cases ----------");

  // Test Case 1: Element found
  const arr1: number[] = [5, 2, 8, 12, 1, 6];
  const target1: number = 8;
  const result1: number = linearSearch(arr1, target1);
  console.log(`Test 1 - Array: [${arr1}], Target: ${target1}`);
  console.log(
    `Expected: 2, Got: ${result1}, Status: ${result1 === 2 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 2: Element not found
  const arr2: number[] = [1, 3, 5, 7, 9];
  const target2: number = 4;
  const result2: number = linearSearch(arr2, target2);
  console.log(`Test 2 - Array: [${arr2}], Target: ${target2}`);
  console.log(
    `Expected: -1, Got: ${result2}, Status: ${
      result2 === -1 ? "PASS" : "FAIL"
    }\n`
  );

  // Test Case 3: First element
  const arr3: number[] = [10, 20, 30, 40];
  const target3: number = 10;
  const result3: number = linearSearch(arr3, target3);
  console.log(`Test 3 - Array: [${arr3}], Target: ${target3}`);
  console.log(
    `Expected: 0, Got: ${result3}, Status: ${result3 === 0 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 4: Last element
  const arr4: number[] = [15, 25, 35, 45];
  const target4: number = 45;
  const result4: number = linearSearch(arr4, target4);
  console.log(`Test 4 - Array: [${arr4}], Target: ${target4}`);
  console.log(
    `Expected: 3, Got: ${result4}, Status: ${result4 === 3 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 5: Empty array
  const arr5: number[] = [];
  const target5: number = 1;
  const result5: number = linearSearch(arr5, target5);
  console.log(`Test 5 - Array: [${arr5}], Target: ${target5}`);
  console.log(
    `Expected: -1, Got: ${result5}, Status: ${
      result5 === -1 ? "PASS" : "FAIL"
    }\n`
  );

  // Test Case 6: Single element array (found)
  const arr6: number[] = [42];
  const target6: number = 42;
  const result6: number = linearSearch(arr6, target6);
  console.log(`Test 6 - Array: [${arr6}], Target: ${target6}`);
  console.log(
    `Expected: 0, Got: ${result6}, Status: ${result6 === 0 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 7: Duplicate elements
  const arr7: number[] = [1, 2, 3, 2, 4];
  const target7: number = 2;
  const result7: number = linearSearch(arr7, target7);
  console.log(`Test 7 - Array: [${arr7}], Target: ${target7}`);
  console.log(
    `First occurrence - Expected: 1, Got: ${result7}, Status: ${
      result7 === 1 ? "PASS" : "FAIL"
    }`
  );

  // Test Case 8: String elements
  const arr8: string[] = ["apple", "banana", "cherry", "date"];
  const target8: string = "cherry";
  const result8: number = linearSearch(arr8, target8);
  console.log(`Test 8 - Array: [${arr8}], Target: ${target8}`);
  console.log(
    `Expected: 2, Got: ${result8}, Status: ${result8 === 2 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 9: Boolean elements
  const arr9: boolean[] = [true, false, true, false];
  const target9: boolean = false;
  const result9: number = linearSearch(arr9, target9);
  console.log(`Test 9 - Array: [${arr9}], Target: ${target9}`);
  console.log(
    `Expected: 1, Got: ${result9}, Status: ${result9 === 1 ? "PASS" : "FAIL"}\n`
  );

  // Test Case 10: Error handling
  try {
    linearSearch(null as any, 5);
    console.log("Test 10 - Error handling: FAIL (No error thrown)");
  } catch (error) {
    console.log(`Test 10 - Error handling: PASS (${(error as Error).message})`);
  }
}

File: C/algorithms/searching/linear_search.c
/*
 * Algorithm Name:
 * Linear Search (Sequential Search)
 *
 * Programming Language:
 * C
 *
 * Category:
 * Searching
 *
 * Difficulty Level:
 * Easy
 *
 * Algorithm Description:
 *
 * 1. Problem it solves:
 *    -> Linear Search finds the position of a target element in an array by checking
 *       each element sequentially from the beginning until the target is found or
 *       the end of the array is reached.
 *
 * 2. Approach / Idea:
 *    -> Start from the first element of the array
 *    -> Compare each element with the target value
 *    -> If match found, return the index
 *    -> If end of array reached without finding target, return -1
 *    -> Works on both sorted and unsorted arrays
 *
 * 3. Complexity:
 *    Time: O(n) - worst case when element is at end or not present
 *    Space: O(1) - constant extra space
 *
 * 4. Variants Implemented:
 *    -> Basic Linear Search
 *    -> Linear Search with All Occurrences
 *    -> Recursive Linear Search
 *    -> Linear Search from End (Reverse)
 *
 * Author:
 * Rishan Menezes
 */

#include <stdio.h>
#include <stdlib.h>

// Function prototypes
int linearSearch(int arr[], int n, int target);
int* linearSearchAllOccurrences(int arr[], int n, int target, int* count);
int linearSearchRecursive(int arr[], int n, int target, int index);
int linearSearchReverse(int arr[], int n, int target);
void printArray(int arr[], int n);

int main() {
    int choice, n, target, result;
    
    printf("=== Linear Search Algorithm Demo ===\n\n");
    
    // Get array size
    printf("Enter the number of elements: ");
    scanf("%d", &n);
    
    if (n <= 0) {
        printf("Invalid array size!\n");
        return 1;
    }
    
    int arr[n];
    
    // Input array elements
    printf("Enter %d elements: ", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }
    
    printf("\nArray: ");
    printArray(arr, n);
    
    printf("\nEnter the element to search: ");
    scanf("%d", &target);
    
    // Menu for different search variants
    printf("\nChoose search variant:\n");
    printf("1. Basic Linear Search\n");
    printf("2. Find All Occurrences\n");
    printf("3. Recursive Linear Search\n");
    printf("4. Reverse Linear Search\n");
    printf("Enter your choice (1-4): ");
    scanf("%d", &choice);
    
    printf("\n=== Results ===\n");
    
    switch (choice) {
        case 1: {
            result = linearSearch(arr, n, target);
            if (result != -1) {
                printf("Element %d found at index %d (position %d)\n", target, result, result + 1);
            } else {
                printf("Element %d not found in the array\n", target);
            }
            break;
        }
        
        case 2: {
            int count = 0;
            int* indices = linearSearchAllOccurrences(arr, n, target, &count);
            
            if (count > 0) {
                printf("Element %d found %d time(s) at indices: ", target, count);
                for (int i = 0; i < count; i++) {
                    printf("%d", indices[i]);
                    if (i < count - 1) printf(", ");
                }
                printf("\n");
                free(indices);
            } else {
                printf("Element %d not found in the array\n", target);
            }
            break;
        }
        
        case 3: {
            result = linearSearchRecursive(arr, n, target, 0);
            if (result != -1) {
                printf("Element %d found at index %d (position %d) using recursion\n", target, result, result + 1);
            } else {
                printf("Element %d not found in the array using recursion\n", target);
            }
            break;
        }
        
        case 4: {
            result = linearSearchReverse(arr, n, target);
            if (result != -1) {
                printf("Element %d found at index %d (position %d) searching from end\n", target, result, result + 1);
            } else {
                printf("Element %d not found in the array searching from end\n", target);
            }
            break;
        }
        
        default:
            printf("Invalid choice!\n");
            return 1;
    }
    
    return 0;
}

/**
 * Basic Linear Search
 * Searches for target element from beginning to end
 * 
 * @param arr: Array to search in
 * @param n: Size of the array
 * @param target: Element to search for
 * @return: Index of element if found, -1 otherwise
 */
int linearSearch(int arr[], int n, int target) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == target) {
            return i;  // Return index where element is found
        }
    }
    return -1;  // Element not found
}

/**
 * Linear Search - Find All Occurrences
 * Finds all positions where target element appears
 * 
 * @param arr: Array to search in
 * @param n: Size of the array
 * @param target: Element to search for
 * @param count: Pointer to store number of occurrences found
 * @return: Dynamically allocated array of indices where element is found
 */
int* linearSearchAllOccurrences(int arr[], int n, int target, int* count) {
    int* indices = (int*)malloc(n * sizeof(int));  // Worst case: all elements match
    *count = 0;
    
    for (int i = 0; i < n; i++) {
        if (arr[i] == target) {
            indices[*count] = i;
            (*count)++;
        }
    }
    
    if (*count == 0) {
        free(indices);
        return NULL;
    }
    
    // Resize array to actual count to save memory
    indices = (int*)realloc(indices, (*count) * sizeof(int));
    return indices;
}

/**
 * Recursive Linear Search
 * Implements linear search using recursion
 * 
 * @param arr: Array to search in
 * @param n: Size of the array
 * @param target: Element to search for
 * @param index: Current index being checked (starts from 0)
 * @return: Index of element if found, -1 otherwise
 */
int linearSearchRecursive(int arr[], int n, int target, int index) {
    // Base case: reached end of array
    if (index >= n) {
        return -1;
    }
    
    // Base case: element found
    if (arr[index] == target) {
        return index;
    }
    
    // Recursive case: search in remaining array
    return linearSearchRecursive(arr, n, target, index + 1);
}

/**
 * Reverse Linear Search
 * Searches for target element from end to beginning
 * Useful when target is more likely to be at the end
 * 
 * @param arr: Array to search in
 * @param n: Size of the array
 * @param target: Element to search for
 * @return: Index of element if found, -1 otherwise
 */
int linearSearchReverse(int arr[], int n, int target) {
    for (int i = n - 1; i >= 0; i--) {
        if (arr[i] == target) {
            return i;  // Return index where element is found
        }
    }
    return -1;  // Element not found
}

/**
 * Utility function to print array elements
 * 
 * @param arr: Array to print
 * @param n: Size of the array
 */
void printArray(int arr[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", arr[i]);
        if (i < n - 1) {
            printf(", ");
        }
    }
    printf("]\n");
}
